File: packages/ts/spaces/src/resource-requests.ts
/**
 * Sealed resource-request inbox — the generic "request-to-create" pattern.
 *
 * A REQUESTER holds only the owner's **public identity link** (no authority, safe to
 * share openly). They seal a typed resource-creation request to the owner's KEM key
 * and append it anonymously to the owner's existing `inbox/{ownerId}/{shard}` collection.
 * The OWNER's reconciler trial-unseals pending requests, creates the requested node in
 * its space with its own owner cap, and seals a narrow per-node cap back to the requester's inbox.
 *
 * Security:
 *   - Offline binding: `verifyIdentityLinkBinding(ownerLink, session)` before sealing.
 *   - Sender authenticity: `sealed.entry.addedBy === req.requester.edPub`.
 *   - Accept/reject gate: owner decides; nothing lands in the space automatically.
 *   - Idempotency: nodes carry `meta.reqId`; `scanResourceRequests` skips fulfilled reqIds.
 */
import type { SealedBlob } from "./config.js"
import { sealToRecipient, unsealFromRecipient } from "./account-seal.js"
import { inboxShard, inboxShards, pullInbox } from "./inbox.js"
import { createKeyedStore } from "./keyed-store.js"
import { verifyIdentityLinkBinding, verifyIdentityLinkKeys } from "./identity-link.js"
import type { IdentityLink } from "./identity-link.js"
import { createNode, inviteToNode, acceptNodeInvite } from "./nodes.js"
import { verifyKemSig, signKemSig } from "./request-verify.js"
import { readObjectTree } from "./object-index.js"
import { randomId } from "@drakkar.software/starfish-protocol"
import type { Session } from "./session.js"
import type { ObjectNode } from "./config.js"
import type { ResourceRequest, ResourceGrant, ResourceReject } from "./token-types.js"
import { makeAnonSpaceClient } from "./client.js"

export type { ResourceRequest, ResourceGrant, ResourceReject }

/**
 * AES-GCM additional-data context binding for inbox seals.
 * Parameterized by `session.inboxAadNamespace` instead of hardcoded namespace.
 */
const inboxAad = (session: Session, recipientId: string, shard: string, kind: string) =>
  `${session.inboxAadNamespace}:${recipientId}:${shard}:${kind}`

/**
 * Trial-unseal an inbox element for `session.userId` with the kind-bound AAD.
 * Returns the plaintext, or `null` when the element isn't sealed to us.
 */
async function tryUnsealInbox(
  session: Session,
  sealed: SealedBlob,
  shard: string,
  mkind: string | undefined,
  defaultKind: string,
): Promise<string | null> {
  try {
    return await unsealFromRecipient(session, sealed, inboxAad(session, session.userId, shard, mkind ?? defaultKind))
  } catch {
    return null
  }
}

/**
 * Seal `obj` to a recipient and append it to their current-shard inbox.
 */
async function sealAppend(
  session: Session,
  recipientUserId: string,
  recipientKemPub: string,
  kind: string,
  obj: unknown,
): Promise<void> {
  const shard = inboxShard()
  const sealed = await sealToRecipient(
    session,
    recipientKemPub,
    JSON.stringify(obj),
    inboxAad(session, recipientUserId, shard, kind),
  )
  const element = { sealed, ts: Date.now(), mkind: kind } as unknown as Record<string, unknown>
  // Use an anonymous client for public-write inbox appends.
  const anonClient = makeAnonSpaceClient({ baseUrl: session.baseUrl, namespace: session.namespace })
  await anonClient.appendAnonymous(
    session.layout.inboxPush(recipientUserId, shard),
    element,
    { edPubHex: session.keys.edPub, edPrivHex: session.keys.edPriv },
  )
}

/**
 * Generic inbox scan: walk both shards, pull each element, trial-unseal it,
 * JSON-parse, and hand the parsed payload + its sealed envelope to `handle`.
 */
async function scanInbox(
  session: Session,
  defaultKind: string,
  handle: (parsed: unknown, sealed: SealedBlob) => void | Promise<void>,
): Promise<void> {
  for (const shard of inboxShards()) {
    const items = await pullInbox(session.accountClient, session.userId, shard, session)
    for (const item of items) {
      const payload = item?.data as Partial<InboxPayload> | undefined
      if (!payload?.sealed) continue

      const plaintext = await tryUnsealInbox(session, payload.sealed, shard, payload.mkind, defaultKind)
      if (plaintext === null) continue

      let parsed: unknown
      try {
        parsed = JSON.parse(plaintext)
      } catch {
        continue
      }
      await handle(parsed, payload.sealed)
    }
  }
}

// ── Owner-store: reqId → owner edPub ─────────────────────────────────────────

const reqIdOwnerStore = createKeyedStore<string>()

/** Record the owner edPub for a submitted request. */
export const saveReqIdOwner = (reqId: string, ownerEdPub: string): void => reqIdOwnerStore.set(reqId, ownerEdPub)
/** Snapshot the store for persistence across reloads. */
export const serializeReqIdOwnerStore = reqIdOwnerStore.serialize
/** Restore the store after a reload (additive). */
export const hydrateReqIdOwnerStore = reqIdOwnerStore.hydrate
/** Clear the store (e.g. on sign-out). */
export const clearReqIdOwnerStore = reqIdOwnerStore.clear

// ── Payload type ──────────────────────────────────────────────────────────────

interface InboxPayload {
  sealed: SealedBlob
  ts: number
  mkind?: string
}

// ── REQUESTER: submit a request ───────────────────────────────────────────────

/** Options for {@link submitResourceRequest}. */
export interface SubmitResourceRequestOptions {
  spaceId: string
  nodeType: string
  title: string
  meta?: Record<string, unknown>
  message?: string
}

/**
 * REQUESTER: send a sealed resource-creation request to an owner's inbox.
 *
 * Returns the `reqId` generated for this request — save it to track fulfilment.
 */
export async function submitResourceRequest(
  session: Session,
  ownerLink: IdentityLink,
  opts: SubmitResourceRequestOptions,
): Promise<{ reqId: string }> {
  if (ownerLink.ownerId === session.userId) throw new Error("Cannot send a request to yourself.")

  if (!(await verifyIdentityLinkBinding(ownerLink, session))) {
    throw new Error("That identity link is malformed — ownerId does not match edPub.")
  }
  await verifyIdentityLinkKeys(ownerLink, session)

  const reqId = randomId()
  const request: ResourceRequest = {
    v: 1,
    kind: "create-resource",
    reqId,
    spaceId: opts.spaceId,
    nodeType: opts.nodeType,
    title: opts.title,
    ...(opts.meta ? { meta: opts.meta } : {}),
    ...(opts.message ? { message: opts.message } : {}),
    requester: {
      userId: session.userId,
      edPub: session.keys.edPub,
      kemPub: session.keys.kemPub,
      kemSig: signKemSig(session.keys),
    },
  }

  saveReqIdOwner(reqId, ownerLink.edPub)
  await sealAppend(session, ownerLink.ownerId, ownerLink.kemPub, "request", request)

  return { reqId }
}

// ── OWNER: scan pending requests ──────────────────────────────────────────────

/** A pending request returned by {@link scanResourceRequests}, not yet fulfilled. */
export interface PendingRequest {
  req: ResourceRequest
  senderEdPub: string
}

/**
 * OWNER: scan this session's inbox for pending `create-resource` requests.
 */
export async function scanResourceRequests(
  session: Session,
  spaceIds?: ReadonlySet<string>,
): Promise<PendingRequest[]> {
  const treeCache = new Map<string, ObjectNode[]>()

  const out: PendingRequest[] = []
  await scanInbox(session, "request", async (parsed, sealed) => {
    const req = parsed as Partial<ResourceRequest>

    if (
      req.v !== 1 ||
      req.kind !== "create-resource" ||
      typeof req.reqId !== "string" ||
      typeof req.spaceId !== "string" ||
      typeof req.nodeType !== "string" ||
      typeof req.title !== "string" ||
      !req.requester ||
      typeof req.requester.edPub !== "string" ||
      typeof req.requester.kemPub !== "string" ||
      typeof req.requester.userId !== "string" ||
      typeof req.requester.kemSig !== "string"
    ) {
      return
    }

    if (sealed.entry.addedBy !== req.requester.edPub) return

    if ((await session.userIdFromEdPub(req.requester.edPub)) !== req.requester.userId) return

    if (!verifyKemSig(req.requester.edPub, req.requester.kemPub, req.requester.kemSig)) return

    if (spaceIds && !spaceIds.has(req.spaceId)) return

    if (!treeCache.has(req.spaceId)) {
      const tree = await readObjectTree(session, req.spaceId).catch(() => [])
      treeCache.set(req.spaceId, tree)
    }
    const tree = treeCache.get(req.spaceId) ?? []
    const alreadyFulfilled = tree.some(
      (n) => (n.meta as Record<string, unknown> | undefined)?.reqId === req.reqId,
    )
    if (alreadyFulfilled) return

    out.push({ req: req as ResourceRequest, senderEdPub: sealed.entry.addedBy })
  })
  return out
}

// ── OWNER: accept a request ───────────────────────────────────────────────────

/** Return value of an accepted {@link acceptResourceRequest}. */
export interface AcceptResult {
  spaceId: string
  nodeId: string
}

/**
 * OWNER: accept a pending resource request — create the node and grant the requester
 * a narrow per-node cap sealed back to their inbox.
 */
export async function acceptResourceRequest(
  session: Session,
  pending: PendingRequest,
  opts?: {
    create?: (session: Session, req: ResourceRequest) => Promise<{ nodeId: string }>
    write?: boolean
    enc?: boolean
  },
): Promise<AcceptResult> {
  const { req } = pending

  let nodeId: string
  if (opts?.create) {
    ;({ nodeId } = await opts.create(session, req))
  } else {
    const node = await createNode(session, req.spaceId, {
      type: req.nodeType,
      title: req.title,
      meta: { ...(req.meta ?? {}), reqId: req.reqId },
      access: "invite",
      enc: opts?.enc ?? false,
    })
    nodeId = node.id
  }

  const bundleJson = await inviteToNode(
    session,
    req.spaceId,
    nodeId,
    JSON.stringify(req.requester),
    { enc: opts?.enc ?? false },
    req.title,
    { isolated: true, write: opts?.write ?? true },
  )

  const grant: ResourceGrant = {
    v: 1,
    kind: "grant",
    reqId: req.reqId,
    spaceId: req.spaceId,
    nodeId,
    bundle: bundleJson,
  }
  await sealAppend(session, req.requester.userId, req.requester.kemPub, "grant", grant)

  return { spaceId: req.spaceId, nodeId }
}

// ── OWNER: reject a request ───────────────────────────────────────────────────

/**
 * OWNER: reject a pending request.
 */
export async function rejectResourceRequest(
  session: Session,
  pending: PendingRequest,
  reason?: string,
): Promise<void> {
  const { req } = pending
  const rejection: ResourceReject = {
    v: 1,
    kind: "reject",
    reqId: req.reqId,
    ...(reason ? { reason } : {}),
  }
  await sealAppend(session, req.requester.userId, req.requester.kemPub, "reject", rejection)
}

// ── REQUESTER: scan grants ────────────────────────────────────────────────────

/**
 * REQUESTER: scan this session's own inbox for resource grants (accepted requests).
 */
export async function scanResourceGrants(
  session: Session,
  opts?: { seenReqIds?: Set<string> },
): Promise<ResourceGrant[]> {
  const out: ResourceGrant[] = []
  const seenReqIds = opts?.seenReqIds ?? new Set<string>()
  await scanInbox(session, "grant", (parsed, sealed) => {
    const msg = parsed as Partial<ResourceGrant | ResourceReject>

    if (msg.v !== 1 || msg.kind !== "grant") return
    const g = msg as Partial<ResourceGrant>
    if (
      typeof g.reqId !== "string" ||
      typeof g.spaceId !== "string" ||
      typeof g.nodeId !== "string" ||
      typeof g.bundle !== "string"
    ) {
      return
    }

    const expectedOwnerEdPub = reqIdOwnerStore.get(g.reqId!)
    if (expectedOwnerEdPub && sealed.entry.addedBy !== expectedOwnerEdPub) return

    if (seenReqIds.has(g.reqId!)) return
    seenReqIds.add(g.reqId!)
    out.push(g as ResourceGrant)
  })
  return out
}

// ── REQUESTER: scan rejects ───────────────────────────────────────────────────

/**
 * REQUESTER: scan this session's own inbox for resource rejections.
 */
export async function scanResourceRejects(
  session: Session,
  opts?: { seenReqIds?: Set<string> },
): Promise<ResourceReject[]> {
  const out: ResourceReject[] = []
  const seenReqIds = opts?.seenReqIds ?? new Set<string>()
  await scanInbox(session, "reject", (parsed, sealed) => {
    const msg = parsed as Partial<ResourceReject>

    if (msg.v !== 1 || msg.kind !== "reject" || typeof msg.reqId !== "string") return

    const expectedOwnerEdPub = reqIdOwnerStore.get(msg.reqId)
    if (expectedOwnerEdPub && sealed.entry.addedBy !== expectedOwnerEdPub) return

    if (seenReqIds.has(msg.reqId)) return
    seenReqIds.add(msg.reqId)
    out.push(msg as ResourceReject)
  })
  return out
}

// ── REQUESTER: accept a grant ─────────────────────────────────────────────────

/**
 * REQUESTER: accept a resource grant — store the per-node cap and return the node reference.
 */
export async function acceptResourceGrant(
  session: Session,
  grant: ResourceGrant,
): Promise<{ spaceId: string; nodeId: string }> {
  const nodeId = await acceptNodeInvite(session, grant.bundle)
  return { spaceId: grant.spaceId, nodeId }
}

File: packages/python/spaces/starfish_spaces/resource_requests.py
"""Sealed resource-request inbox — the generic "request-to-create" pattern.

A REQUESTER holds only the owner's **public identity link** (no authority, safe
to share openly). They seal a typed resource-creation request to the owner's KEM
key and append it anonymously to the owner's existing ``inbox/{ownerId}/{shard}``
collection. The OWNER's reconciler trial-unseals pending requests, creates the
requested node in its space with its own owner cap, and seals a narrow per-node
cap back to the requester's inbox.

Security:
- Offline binding: ``verify_identity_link_binding(owner_link, session)`` before sealing.
- Sender authenticity: ``sealed.entry.addedBy === req.requester.edPub``.
- Accept/reject gate: owner decides; nothing lands in the space automatically.
- Idempotency: nodes carry ``meta.reqId``; ``scan_resource_requests`` skips
  fulfilled reqIds.
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any, Callable, Optional

from starfish_protocol.random import random_id

from starfish_spaces.account_seal import seal_to_recipient, unseal_from_recipient
from starfish_spaces.identity_link import (
    IdentityLink,
    verify_identity_link_binding,
    verify_identity_link_keys,
)
from starfish_spaces.inbox import inbox_shard, inbox_shards, pull_inbox
from starfish_spaces.keyed_store import create_keyed_store
from starfish_spaces.nodes import accept_node_invite, create_node, invite_to_node
from starfish_spaces.object_index import read_object_tree
from starfish_spaces.request_verify import sign_kem_sig, verify_kem_sig
from starfish_spaces.token_types import ResourceGrant, ResourceReject, ResourceRequest

if TYPE_CHECKING:
    from starfish_spaces.session import Session


# ── AAD helpers ───────────────────────────────────────────────────────────────


def _inbox_aad(session: "Session", recipient_id: str, shard: str, kind: str) -> str:
    return f"{session.inbox_aad_namespace}:{recipient_id}:{shard}:{kind}"


async def _try_unseal_inbox(
    session: "Session",
    sealed: dict[str, Any],
    shard: str,
    mkind: Optional[str],
    default_kind: str,
) -> Optional[str]:
    try:
        aad = _inbox_aad(session, session.user_id, shard, mkind or default_kind)
        return unseal_from_recipient(session.keys, sealed, aad)
    except Exception:
        return None


# ── Append sealed message to a recipient's inbox ──────────────────────────────


async def _seal_append(
    session: "Session",
    recipient_user_id: str,
    recipient_kem_pub: str,
    kind: str,
    obj: Any,
) -> None:
    """Seal ``obj`` to ``recipient_kem_pub`` and append it to the current-shard inbox."""
    import time as _time
    from starfish_spaces.client import make_anon_space_client, ClientOpts

    shard = inbox_shard()
    aad = _inbox_aad(session, recipient_user_id, shard, kind)
    sealed = seal_to_recipient(session.keys, "", recipient_kem_pub, json.dumps(obj), aad)

    element: dict[str, Any] = {
        "sealed": sealed,
        "ts": int(_time.time() * 1000),
        "mkind": kind,
    }

    anon_opts: ClientOpts = {"baseUrl": session.base_url, "namespace": session.namespace}  # type: ignore[misc]
    anon_client = make_anon_space_client(anon_opts)
    push_path = session.layout.inbox_push(recipient_user_id, shard)
    await anon_client.append_anonymous(push_path, element)


# ── Generic inbox scan ────────────────────────────────────────────────────────


async def _scan_inbox(
    session: "Session",
    default_kind: str,
    handle: Callable[[Any, dict[str, Any]], Any],
) -> None:
    """Walk both current + previous shards, trial-unseal each element, call ``handle``."""
    for shard in inbox_shards():
        items = await pull_inbox(session.account_client, session, session.user_id, shard)
        for item in items:
            payload = item.get("data") if isinstance(item, dict) else None
            if not isinstance(payload, dict) or not payload.get("sealed"):
                continue
            sealed = payload["sealed"]
            plaintext = await _try_unseal_inbox(session, sealed, shard, payload.get("mkind"), default_kind)
            if plaintext is None:
                continue
            try:
                parsed = json.loads(plaintext)
            except Exception:
                continue
            result = handle(parsed, sealed)
            if hasattr(result, "__await__"):
                await result


# ── Owner-store: reqId → owner edPub ─────────────────────────────────────────

_req_id_owner_store = create_keyed_store()


def save_req_id_owner(req_id: str, owner_ed_pub: str) -> None:
    """Record the owner edPub for a submitted request."""
    _req_id_owner_store.set(req_id, owner_ed_pub)


def serialize_req_id_owner_store() -> str:
    return _req_id_owner_store.serialize()


def hydrate_req_id_owner_store(raw: str) -> None:
    _req_id_owner_store.hydrate(raw)


def clear_req_id_owner_store() -> None:
    _req_id_owner_store.clear_all()


# ── REQUESTER: submit a request ───────────────────────────────────────────────


async def submit_resource_request(
    session: "Session",
    owner_link: IdentityLink,
    opts: dict[str, Any],
) -> dict[str, str]:
    """REQUESTER: send a sealed resource-creation request to an owner's inbox.

    Args:
        session:    The requester's session.
        owner_link: The owner's decoded :class:`IdentityLink`.
        opts:       ``{spaceId, nodeType, title, meta?, message?}``.

    Returns:
        ``{"reqId": str}`` — save it to track fulfilment.

    Raises:
        ValueError: if sending to yourself or if the owner link is malformed.
    """
    if owner_link.owner_id == session.user_id:
        raise ValueError("Cannot send a request to yourself.")
    if not await verify_identity_link_binding(owner_link, session):
        raise ValueError("That identity link is malformed — ownerId does not match edPub.")
    await verify_identity_link_keys(owner_link, session)

    req_id = random_id()
    kem_sig = sign_kem_sig(session.keys["kemPub"], session.keys["edPriv"])
    request: ResourceRequest = {  # type: ignore[assignment]
        "v": 1,
        "kind": "create-resource",
        "reqId": req_id,
        "spaceId": opts["spaceId"],
        "nodeType": opts["nodeType"],
        "title": opts["title"],
        **({"meta": opts["meta"]} if opts.get("meta") else {}),
        **({"message": opts["message"]} if opts.get("message") else {}),
        "requester": {
            "userId": session.user_id,
            "edPub": session.keys["edPub"],
            "kemPub": session.keys["kemPub"],
            "kemSig": kem_sig,
        },
    }

    save_req_id_owner(req_id, owner_link.ed_pub)
    await _seal_append(session, owner_link.owner_id, owner_link.kem_pub, "request", request)
    return {"reqId": req_id}


# ── OWNER: scan pending requests ──────────────────────────────────────────────


class PendingRequest(dict):
    """A pending request returned by :func:`scan_resource_requests`."""
    pass


async def scan_resource_requests(
    session: "Session",
    space_ids: Optional[frozenset[str]] = None,
) -> list[dict[str, Any]]:
    """OWNER: scan this session's inbox for pending ``create-resource`` requests.

    Returns:
        List of ``{"req": ResourceRequest, "senderEdPub": str}`` dicts.
    """
    tree_cache: dict[str, list[Any]] = {}
    out: list[dict[str, Any]] = []

    async def handle(parsed: Any, sealed: dict[str, Any]) -> None:
        req = parsed if isinstance(parsed, dict) else {}

        if (
            req.get("v") != 1
            or req.get("kind") != "create-resource"
            or not isinstance(req.get("reqId"), str)
            or not isinstance(req.get("spaceId"), str)
            or not isinstance(req.get("nodeType"), str)
            or not isinstance(req.get("title"), str)
            or not isinstance(req.get("requester"), dict)
        ):
            return

        requester = req["requester"]
        if not (
            isinstance(requester.get("edPub"), str)
            and isinstance(requester.get("kemPub"), str)
            and isinstance(requester.get("userId"), str)
            and isinstance(requester.get("kemSig"), str)
        ):
            return

        # Verify sender authenticity from the sealed entry's addedBy field.
        # Fail-closed: a missing/empty addedBy is treated as a mismatch.
        added_by = sealed.get("addedBy") or (sealed.get("entry", {}) or {}).get("addedBy")
        if added_by != requester["edPub"]:
            return

        if (await session.user_id_from_ed_pub(requester["edPub"])) != requester["userId"]:
            return

        if not verify_kem_sig(requester["edPub"], requester["kemPub"], requester["kemSig"]):
            return

        if space_ids is not None and req["spaceId"] not in space_ids:
            return

        # Check if already fulfilled (node with meta.reqId exists in space).
        space_id = req["spaceId"]
        if space_id not in tree_cache:
            try:
                tree = await read_object_tree(session.content_client, session, space_id)
                tree_cache[space_id] = tree
            except Exception:
                tree_cache[space_id] = []

        already_fulfilled = any(
            (getattr(n, "meta", None) or {}).get("reqId") == req["reqId"]
            for n in tree_cache[space_id]
        )
        if already_fulfilled:
            return

        out.append({"req": req, "senderEdPub": requester["edPub"]})

    await _scan_inbox(session, "request", handle)
    return out


# ── OWNER: accept a request ───────────────────────────────────────────────────


async def accept_resource_request(
    session: "Session",
    pending: dict[str, Any],
    opts: Optional[dict[str, Any]] = None,
) -> dict[str, str]:
    """OWNER: accept a pending resource request.

    Creates the node (via ``opts.create`` or the default :func:`create_node`),
    grants the requester a narrow per-node cap sealed back to their inbox.

    Returns:
        ``{"spaceId": str, "nodeId": str}``.
    """
    req: dict[str, Any] = pending["req"]
    create_fn = (opts or {}).get("create")

    if create_fn:
        result = await create_fn(session, req)
        node_id = result["nodeId"]
    else:
        node = await create_node(session, req["spaceId"], {
            "type": req["nodeType"],
            "title": req["title"],
            "meta": {**(req.get("meta") or {}), "reqId": req["reqId"]},
            "access": "invite",
            "enc": bool((opts or {}).get("enc")),
        })
        node_id = node.id

    requester = req["requester"]
    bundle_json = await invite_to_node(
        session, req["spaceId"], node_id,
        json.dumps(requester),
        {"enc": bool((opts or {}).get("enc"))},
        req["title"],
        {"isolated": True, "write": (opts or {}).get("write", True)},
    )

    grant: ResourceGrant = {  # type: ignore[assignment]
        "v": 1,
        "kind": "grant",
        "reqId": req["reqId"],
        "spaceId": req["spaceId"],
        "nodeId": node_id,
        "bundle": bundle_json,
    }
    await _seal_append(session, requester["userId"], requester["kemPub"], "grant", grant)
    return {"spaceId": req["spaceId"], "nodeId": node_id}


# ── OWNER: reject a request ───────────────────────────────────────────────────


async def reject_resource_request(
    session: "Session",
    pending: dict[str, Any],
    reason: Optional[str] = None,
) -> None:
    """OWNER: reject a pending request."""
    req = pending["req"]
    requester = req["requester"]
    rejection: ResourceReject = {  # type: ignore[assignment]
        "v": 1,
        "kind": "reject",
        "reqId": req["reqId"],
        **({"reason": reason} if reason else {}),
    }
    await _seal_append(session, requester["userId"], requester["kemPub"], "reject", rejection)


# ── REQUESTER: scan grants ────────────────────────────────────────────────────


async def scan_resource_grants(
    session: "Session",
    seen_req_ids: Optional[set[str]] = None,
) -> list[ResourceGrant]:
    """REQUESTER: scan this session's inbox for resource grants."""
    out: list[ResourceGrant] = []
    seen = seen_req_ids if seen_req_ids is not None else set()

    def handle(parsed: Any, sealed: dict[str, Any]) -> None:
        msg = parsed if isinstance(parsed, dict) else {}
        if msg.get("v") != 1 or msg.get("kind") != "grant":
            return
        if not (isinstance(msg.get("reqId"), str) and isinstance(msg.get("spaceId"), str)
                and isinstance(msg.get("nodeId"), str) and isinstance(msg.get("bundle"), str)):
            return

        req_id = msg["reqId"]
        expected_owner = _req_id_owner_store.get(req_id)
        added_by = sealed.get("addedBy") or (sealed.get("entry", {}) or {}).get("addedBy")
        if expected_owner and added_by != expected_owner:
            return
        if req_id in seen:
            return
        seen.add(req_id)
        out.append(msg)  # type: ignore[arg-type]

    await _scan_inbox(session, "grant", handle)
    return out


# ── REQUESTER: scan rejects ───────────────────────────────────────────────────


async def scan_resource_rejects(
    session: "Session",
    seen_req_ids: Optional[set[str]] = None,
) -> list[ResourceReject]:
    """REQUESTER: scan this session's inbox for resource rejections."""
    out: list[ResourceReject] = []
    seen = seen_req_ids if seen_req_ids is not None else set()

    def handle(parsed: Any, sealed: dict[str, Any]) -> None:
        msg = parsed if isinstance(parsed, dict) else {}
        if msg.get("v") != 1 or msg.get("kind") != "reject" or not isinstance(msg.get("reqId"), str):
            return
        req_id = msg["reqId"]
        expected_owner = _req_id_owner_store.get(req_id)
        added_by = sealed.get("addedBy") or (sealed.get("entry", {}) or {}).get("addedBy")
        if expected_owner and added_by != expected_owner:
            return
        if req_id in seen:
            return
        seen.add(req_id)
        out.append(msg)  # type: ignore[arg-type]

    await _scan_inbox(session, "reject", handle)
    return out


# ── REQUESTER: accept a grant ─────────────────────────────────────────────────


async def accept_resource_grant(
    session: "Session",
    grant: ResourceGrant,
) -> dict[str, str]:
    """REQUESTER: store the per-node cap and return the node reference."""
    node_id = await accept_node_invite(session, grant["bundle"])
    return {"spaceId": grant["spaceId"], "nodeId": node_id}


__all__ = [
    "PendingRequest",
    "save_req_id_owner",
    "serialize_req_id_owner_store",
    "hydrate_req_id_owner_store",
    "clear_req_id_owner_store",
    "submit_resource_request",
    "scan_resource_requests",
    "accept_resource_request",
    "reject_resource_request",
    "scan_resource_grants",
    "scan_resource_rejects",
    "accept_resource_grant",
]
